README.md
<div align="center">

# Musd

[中文说明](README.zh-CN.md)

[![Release](https://github.com/hustcer/musd/actions/workflows/release.yaml/badge.svg)](https://github.com/hustcer/musd/actions/workflows/release.yaml)

🎵 A Rust CLI App to download super high quality musics🎵

![](https://img.alicdn.com/imgextra/i1/O1CN01oWvzdy1xIPUyZEyAK_!!6000000006420-1-tps-1964-878.gif)

</div>

`musd` is a MUSic Downloader, and that's why it was named as `musd`.

## Installation

### Install from binary release

You can download the binary tar ball according to your OS from the [Release Page](https://github.com/hustcer/musd/releases), and run the `musd` executable file directly.

### Install from source

You need to have `cargo` been installed, if you haven't installed it yet here is a [guide](https://www.rust-lang.org/tools/install) for you. (Rust v1.60.0 or above required)

```bash
# Install from crates.io
cargo install musd
# OR clone the source code and install it from local disk
git clone git@github.com:hustcer/musd.git
cd musd && cargo install --path .
```

## Usage

```bash
# Search music by name
musd someone like you
# Or search music by singer name
musd Celine Dion
```

`musd` will download `*.flac` format music files by default, if you want `*.m4a` format please specify it by `-f` or `--format` flag. Currently, only two formats available: 'flac' or 'm4a'.

### Save to a specified folder

1. Use `--output` or `-o` to specify the directory that downloaded musics will be saved;
2. Set `MUSD_OUTPUT` env variable to the path where you want to have your musics been saved;

## Resource

- All the musics will be downloaded from [MIGU](https://music.migu.cn/)

## PS

By default, the highest quality lossless music will be downloaded, usually in `flac` format.

This app was heavily inspired by [musicn](https://github.com/zonemeen/musicn), special thanks to them.
`musicn` was written in Node.js and I rewrote it by rust just for practice. And this is my first Rust App.
Hope you love it.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.2.5] - 2022-06-06

### Bug Fixes

- Fix build info check for release packages

### Features

- Bump version v0.2.5 and use static-link-openssl for linux
- Add release-check workflow to test musd binary
- Add build info check for package release

### Miscellaneous Tasks

- Remove cargo* from release package

## [0.2.3] - 2022-05-25

### Features

- Try to do the Github release task by nushell
- Use nushell instead of bash to do the release task
- Add windows msi release package

### Miscellaneous Tasks

- Improve release-pkg script

## [0.2.1] - 2022-05-24

### Features

- Add download m4a format music file support
- Add x86_64-unknown-linux-gnu release target
- Add aarch64-unknown-linux-gnu binary release
- Add armv7-unknown-linux-gnueabihf binary release

## [0.2.0] - 2022-05-21

### Bug Fixes

- Update badge for release.yml

### Documentation

- Add rust min version tip

### Features

- Update github release workflow and enable unknown-linux-musl release
- Add some basic tests
- Add build info table by `-b` option, fix #15 (#18)

### Miscellaneous Tasks

- Remove bad release targets

## [0.1.2] - 2022-05-19

### Bug Fixes

- Fix github release workflow

### Documentation

- Add README.zh-CN.md
- Add build status badge for README.md

### Features

- Add progress bar for downloading status
- Add customizable output path by using `MUSD_OUTPUT` ENV var support
- Add customizable output path by using `--output(-o)` option support
- Add arg_required_else_help support
- Add ci.yaml workflow
- Add github release workflow
- Add usage gif to README.md

### Miscellaneous Tasks

- Update actions for ci.yaml

### Refactor

- Extract configuration constants

### Opt

- Update release config for optimization

## [0.1.0] - 2022-05-17

### Bug Fixes

- Fix music file downloading corruption issue

### Features

- Add search and select music for downloading feature
- Do some code refactor and extract lib.rs
- Add download music feature
- Rename downloaded music by music name and singer
- Highlight output and add no search result tip
- Warn the user if music file already exists
- Add develop related tasks managed by just
- Add CHANGELOG.md

### Miscellaneous Tasks

- Fix rust clippy warnings
- Add LICENSE and empty README.md file



tests/cli.rs
use assert_cmd::Command;
use predicates::prelude::*;
use std::error::Error;

type TestResult = Result<(), Box<dyn Error>>;

#[test]
fn runs_empty() {
    let mut cmd = Command::cargo_bin("musd").unwrap();
    cmd.assert()
        .failure()
        .stderr(predicate::str::contains("USAGE:\n"));
}

#[test]
fn runs_help() -> TestResult {
    Command::cargo_bin("musd")?
        .args(&["-h"])
        .assert()
        .success()
        .stdout(predicate::str::contains("USAGE:\n"));
    Ok(())
}

#[test]
fn test_search() -> TestResult {
    let songs = musd::search("someone like you").unwrap();
    assert!(songs.len() > 5);
    Ok(())
}


src/lib.rs
pub mod build_info;
/**
 * Author: hustcer
 * Created: 2022/05/15 16:52:00
 * Description: A CLI App to search and download musics
 * REFS:
 *  1. https://serde.rs/
 *  2. https://rust-lang-nursery.github.io/rust-cookbook/
 *  3. The map trait on serde_json::Value: https://github.com/serde-rs/json/issues/484
 *  4. Iterating through an serde_json::Value: https://github.com/serde-rs/json/issues/417
 *  5. https://pd.musicapp.migu.cn/MIGUM3.0/v1.0/content/search_all.do?text=传奇&pageSize=1&searchSwitch={song:1}
 *  6. https://rust-lang-nursery.github.io/rust-cookbook/web/clients/download.html
 *  7. https://docs.rs/url/latest/url/index.html
 *  8. https://github.com/mihaigalos/tutorials/tree/master/rust/download_with_progressbar
 *  9. https://www.sitepoint.com/rust-global-variables/
 *
 * TODO:
 *  [x] Search music;
 *  [x] Notify user if there is no matching music；
 *  [x] List music searching results, include name, singer and size, etc.;
 *  [x] Select the music(s) to be downloaded;
 *  [x] Download the selected music;
 *  [x] Stop downloading if music file already exists；
 *  [x] Show progress bar while downloading;
 *  [x] Customizable output path by using ENV var;
 *  [x] Customizable output path by using `--output(-o)` option;
 */
mod def;
mod download;

pub use def::*;

use dialoguer::{theme::ColorfulTheme, Select};
use serde::Deserialize;
use serde_json::Value;
use url::Url;
use yansi::Paint;

/**
 * Search music by music or singer name
 */
#[tokio::main]
pub async fn search(search: &str) -> MusdResult<Vec<Song>> {
    let mut url = Url::parse(MIGU_QUERY_API)?;
    url.query_pairs_mut()
        .append_pair("text", search)
        .append_pair("pageSize", "1")
        .append_pair("searchSwitch", "{song:1}");
    // println!("Current search url {}", url.as_str());
    let resp = reqwest::get(url.as_str()).await?.text().await?;
    let val: Value = serde_json::from_str(&resp)?;
    let result = &val["songResultData"]["result"];

    // Handling no searching result cases
    #[allow(clippy::cmp_owned)]
    if result.to_string() == "null" {
        println!(
            "Can not find {} related musics! Bye ...",
            Paint::green(search).bold()
        );
        std::process::exit(0);
    }
    // let songs = val["songResultData"]["result"].as_array().unwrap();
    let songs = Vec::<Song>::deserialize(result)?;
    // println!("{:#?}", songs);
    Ok(songs)
}

/**
 * Show music selector and press `enter` or `space` to download it
 */
pub fn download_selected(songs: Vec<Song>, args: &Args) -> MusdResult<()> {
    // Filter songs that have super quality
    let sq_songs = songs
        .iter()
        .filter(|s| s.new_rate_formats.iter().any(|f| f.format_type == "SQ"))
        .collect::<Vec<_>>();

    if sq_songs.is_empty() {
        println!(
            "Can not find any lossless music related to {}, Bye ...",
            Paint::green(&args.music.join(" ")).bold()
        );
        std::process::exit(0);
    }

    // Select one music a time, we can use multiple select later
    let selection = select_music(sq_songs.clone());

    // println!("You select {:#?}", sq_songs[selection]);
    // println!("Start to download {}!", Paint::green(&selections[selection]));

    if let Err(e) = download::download_music(sq_songs[selection], args) {
        eprintln!("[ERROR]: {}", Paint::red(e));
        std::process::exit(2);
    }
    Ok(())
}

/**
 * Show music selector with music name, singer and size included
 */
fn select_music(sq_songs: Vec<&Song>) -> usize {
    // Prepare music list for user selecting
    let selections = sq_songs
        .iter()
        .enumerate()
        .map(|(i, s)| {
            let sq_format = s
                .new_rate_formats
                .iter()
                .filter(|f| f.format_type == "SQ")
                .collect::<Vec<_>>();

            // println!("{:#?}", sq_format);
            let idx = i + 1;
            let size = sq_format[0]
                .size
                .parse::<u32>()
                .expect("Music file size parsing error!");
            let size_mb: f32 = (size as f32) / 1024.0 / 1024.0;

            format!("{idx:2}. {} -{} -- {:.2} MB", s.name, s.singers[0], size_mb)
        })
        .collect::<Vec<_>>();

    let prompt = format!(
        "Find {} results with super high quality, Press `enter` or `space` key to download...",
        selections.len()
    );

    // Select one music a time, we can use multiple select later
    Select::with_theme(&ColorfulTheme::default())
        .with_prompt(prompt)
        .default(0)
        .items(&selections[..])
        .interact()
        .expect("Music selection failed!")
}


src/main.rs
/**
 * Author: hustcer
 * Created: 2022/05/15 16:52:00
 * Description: A CLI App to search and download musics
 */
use clap::Parser;
use yansi::Paint;

fn main() {
    // Parsing user input args
    // 'music' values: Some(["someone", "like", "you"])
    // println!("'music' values: {:?}", args.music);
    let args = musd::Args::parse();
    // Disable coloring by `CLICOLOR` env variable
    if let Ok(true) = std::env::var("CLICOLOR").map(|v| v == "0") {
        Paint::disable();
    }

    if args.build_info {
        musd::build_info::show_info();
        std::process::exit(0);
    }

    if let Err(e) =
        musd::search(&args.music.join(" ")).and_then(|songs| musd::download_selected(songs, &args))
    {
        eprintln!("[ERROR]: {}", Paint::red(e));
        std::process::exit(1);
    }
}


src/download.rs
/**
 * Author: hustcer
 * Created: 2022/05/17 13:52:00
 * Description: Music download helper
 */
use crate::def::*;
use futures_util::StreamExt;
use indicatif::{ProgressBar, ProgressStyle};
use std::cmp::min;
use std::fs::File;
use std::io::prelude::*;
use std::path::{Path, PathBuf};
use url::Url;
use yansi::Paint;

/**
 * Download music file and save to local disk
 */
#[tokio::main]
pub async fn download_music(song: &Song, args: &Args) -> MusdResult<()> {
    let dest_dir = get_dest_directory(args)?;
    let download_url = get_download_url(song, args)?;
    let download_src = download_url.to_string();
    // println!("{}", &download_url);
    // Get music file extension from download URI
    let extension = get_file_extension(&download_url);

    let dest_file = format!("{}-{}.{extension}", song.name, song.singers[0].name);
    let dest_path = dest_dir.join(&dest_file);
    // Check file existence, stop downloading if already exists.
    if Path::new(&dest_path).exists() {
        println!(
            "File {} already exists, stop downloading, bye...",
            Paint::green(&dest_path.to_string_lossy())
        );
        std::process::exit(0);
    }

    // Reqwest setup
    let res = reqwest::get(download_url).await?;
    let total_size = res
        .content_length()
        .ok_or_else(|| MusdError::GetLengthFailed(String::from(&download_src)))
        .expect("Get content length failed!");

    println!(
        "Start downloading the music and will be saved to `{}`!",
        Paint::green(&dest_dir.display()).bold()
    );

    // Indicatif setup
    let pb = ProgressBar::new(total_size);
    pb.set_style(ProgressStyle::default_bar()
        .template("{msg}\n{spinner:.green} [{elapsed_precise}] [{wide_bar:.cyan/blue}] {bytes}/{total_bytes} ({bytes_per_sec}, {eta})")
        .expect("Progress bar rendering failed!")
        .progress_chars("# >-")); // "█ >-" OR "#>-"

    let mut downloaded: u64 = 0;
    let mut stream = res.bytes_stream();

    let mut dest = {
        println!("The music to download: {:?}", Paint::green(&dest_file));
        println!("Will be located under: {:?}", Paint::green(&dest_path));
        File::create(dest_path)?
    };
    while let Some(item) = stream.next().await {
        let chunk = item?;
        let _ = dest
            .write_all(&chunk)
            .map_err(|_| "Error while writing to file");
        let new = min(downloaded + (chunk.len() as u64), total_size);
        downloaded = new;
        pb.set_position(new);
    }

    pb.finish_with_message(format!("Downloaded to {}", dest_file));
    Ok(())
}

/**
 * Get music download url
 */
fn get_download_url(song: &Song, args: &Args) -> MusdResult<Url> {
    let sq_format = song
        .new_rate_formats
        .iter()
        .filter(|f| f.format_type == "SQ")
        .collect::<Vec<_>>();

    let music = sq_format
        .get(0)
        .expect("No super quality format available!");

    // println!("{:#?}", &music);
    // Download .flac format music file by default
    let target = if music.android_url.is_empty() {
        &music.url
    } else {
        &music.android_url
    };

    // If user prefer to download .m4a format and ios_url is available
    let target = match args.format {
        Some(MusicType::M4a) if !music.ios_url.is_empty() => &music.ios_url,
        _ => target,
    };
    // Use `https` download instead of `ftp`
    let mut download_url = Url::parse(target)?;
    let result = download_url.set_scheme(MIGU_DOWNLOAD_SCHEME);
    assert!(result.is_ok());
    let result = download_url.set_host(Some(MIGU_DOWNLOAD_HOST));
    assert!(result.is_ok());
    Ok(download_url)
}

/**
 * Get music file extension
 */
fn get_file_extension(download_url: &Url) -> &str {
    let fname = download_url
        .path_segments()
        .and_then(|segments| segments.last())
        .and_then(|name| if name.is_empty() { None } else { Some(name) })
        .unwrap_or("music-tmp.flac");

    Path::new(fname)
        .extension()
        .expect("Get music file extension error!")
        .to_str()
        .expect("Convert music file extension error!")
}

/**
 * Get dest directory to save the downloaded music files
 * 1. Use current directory by default;
 * 2. If `MUSD_OUTPUT` was set and the value as a directory exists, then use it;
 * 3. Finally, `--output(-o)` option will has the hightest priority if it exists;
 */
fn get_dest_directory(args: &Args) -> MusdResult<PathBuf> {
    // println!("{:?}", args);
    let output_env_key = MUSD_OUTPUT_ENV_KEY;
    if let Some(path) = &args.output {
        return Ok(path.to_path_buf());
    }
    // If no `--output(-o)` option specified use current directory by default
    let path = std::env::current_dir()?;
    match std::env::var(output_env_key) {
        Ok(val) if Path::new(&val).exists() => Ok(Path::new(&val).to_path_buf()),
        Err(_) => Ok(path),
        _ => Ok(path),
    }
}


src/build_info.rs
/**
 * Author: hustcer
 * Created: 2022/05/21 10:21:00
 * Description: Show the build info of this App
 */
use shadow_rs::shadow;
use tabled::{object::Segment, style::Style, Alignment, Extract, Modify, TableIteratorExt};

shadow!(build);

pub fn show_info() {
    let is_debug = format!("{}", shadow_rs::is_debug());
    let data = vec![
        ["debug", is_debug.as_str()],
        ["pkg_version", build::PKG_VERSION],
        ["release_tag", build::TAG],
        ["short_commit", build::SHORT_COMMIT],
        ["build_os", build::BUILD_OS],
        ["build_target", build::BUILD_TARGET],
        ["build_time", build::BUILD_TIME],
        ["commit_date", build::COMMIT_DATE],
        ["rust_channel", build::RUST_CHANNEL],
        ["rust_version", build::RUST_VERSION],
    ];

    let table = data
        .table()
        .with(Style::github_markdown()) // psql, blank, github_markdown
        .with(Extract::segment(1.., ..))
        .with(Modify::new(Segment::all()).with(Alignment::left()));

    println!("{}", table);
}


src/def.rs
/**
 * Author: hustcer
 * Created: 2022/05/17 13:52:00
 * Description: Commonly used structs
 */
use clap::{ArgEnum, Parser};
use serde::Deserialize;
use thiserror::Error;

pub type MusdResult<T> = Result<T, MusdError>;

pub const MUSD_OUTPUT_ENV_KEY: &str = "MUSD_OUTPUT";
pub const MIGU_DOWNLOAD_SCHEME: &str = "https";
pub const MIGU_DOWNLOAD_HOST: &str = "freetyst.nf.migu.cn";
pub const MIGU_QUERY_API: &str = "https://pd.musicapp.migu.cn/MIGUM3.0/v1.0/content/search_all.do";

#[derive(Error, Debug)]
pub enum MusdError {
    #[error("Download url parsing error")]
    UrlErr(#[from] url::ParseError),
    #[error("Dest file creation failed")]
    CreationErr(#[from] std::io::Error),
    #[error("Download by reqwest failed")]
    ReqwestErr(#[from] reqwest::Error),
    #[error("Searching response parsing failed")]
    JsonParseErr(#[from] serde_json::Error),
    #[error("Failed to get content length from `{0}`")]
    GetLengthFailed(String),
}

/**
 * REF: https://github.com/clap-rs/clap/blob/v3.1.18/examples/derive_ref/README.md
 */
/// A CLI App to search and download musics
#[derive(Parser, Debug)]
#[clap(version, about, long_about = None, arg_required_else_help(true))]
pub struct Args {
    /// The path to save the downloaded music, current directory by default, and you can use `MUSD_OUTPUT` env to set the default output path too
    #[clap(short, long, parse(from_os_str), value_name = "PATH")]
    pub output: Option<std::path::PathBuf>,

    /// Preferred music format
    #[clap(short, long, arg_enum, value_name = "MUSIC_FORMAT")]
    pub format: Option<MusicType>,

    /// Show the CLI build info detail
    #[clap(short, long)]
    pub build_info: bool,

    /// The music or singer name to search for further downloading
    #[clap(value_name = "MUSIC_OR_SINGER")]
    pub music: Vec<String>,
}

#[derive(ArgEnum, Debug, Clone)]
#[clap(rename_all = "kebab_case")]
pub enum MusicType {
    M4a,
    Flac,
}

#[allow(dead_code)]
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Song {
    pub name: String,
    pub singers: Vec<Singer>,
    pub new_rate_formats: Vec<MusicFormat>,
}

#[allow(dead_code)]
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Singer {
    pub id: String,
    pub name: String,
}

impl std::fmt::Display for Singer {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "({})", self.name)
    }
}

#[allow(dead_code)]
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MusicFormat {
    pub format_type: String,

    #[serde(default = "default_string")]
    pub url: String,

    #[serde(default = "default_string")]
    pub ios_url: String,

    #[serde(default = "default_string")]
    pub android_url: String,

    #[serde(default = "default_size")]
    pub size: String,

    #[serde(default = "default_string")]
    pub file_type: String,
}

fn default_string() -> String {
    "".to_string()
}

fn default_size() -> String {
    "0".to_string()
}
